diff --git b/fs/btrfs/disk-io.c a/fs/btrfs/disk-io.c
index ad72f4e..4354b91 100644
--- b/fs/btrfs/disk-io.c
+++ a/fs/btrfs/disk-io.c
@@ -2788,6 +2788,8 @@ static int barrier_all_devices(struct btrfs_fs_info *info)
 	/* send down all the barriers */
 	head = &info->fs_devices->devices;
 	list_for_each_entry_rcu(dev, head, dev_list) {
+		if (dev->missing)
+			continue;
 		if (!dev->bdev) {
 			errors_send++;
 			continue;
