From ed40564940e759a32e12e2af163b2e18aac7eda2 Mon Sep 17 00:00:00 2001
From: Li Qiang <b19715@freescale.com>
Date: Thu, 24 Apr 2014 09:55:23 +0800
Subject: [PATCH 3/3] iMX6 LCD interface supports 1920x1080i50 mode.

---
 drivers/video/mxc/mxc_lcdif.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/video/mxc/mxc_lcdif.c b/drivers/video/mxc/mxc_lcdif.c
index 5cbdc73..1f83934 100644
--- a/drivers/video/mxc/mxc_lcdif.c
+++ b/drivers/video/mxc/mxc_lcdif.c
@@ -39,6 +39,12 @@ static struct fb_videomode lcdif_modedb[] = {
 	FB_SYNC_CLK_LAT_FALL,
 	FB_VMODE_NONINTERLACED,
 	0,},
+	{
+	/* 1920x1080i @ 50 Hz , pixel clk @ 74.5MHz */
+	"LCD-1080I50", 50, 1920, 1080, 13468, 528, 148, 4, 31, 44, 10,
+	FB_SYNC_HOR_HIGH_ACT | FB_SYNC_VERT_HIGH_ACT,
+	FB_VMODE_INTERLACED,
+	0,},
 };
 static int lcdif_modedb_sz = ARRAY_SIZE(lcdif_modedb);
 
-- 
1.8.0

