diff --git a/include/linux/blk_types.h b/include/linux/blk_types.h
index fa1abeb..f933bc4 100644
--- a/include/linux/blk_types.h
+++ b/include/linux/blk_types.h
@@ -28,6 +28,17 @@ struct bio_vec {
 	unsigned int	bv_offset;
 };
 
+struct bvec_iter {
+	sector_t		bi_sector;      /* device address in 512 byte
+                                                  sectors */
+	unsigned int		bi_size;        /* residual I/O count */
+
+	unsigned int		bi_idx;         /* current index into bvl_vec */
+
+	unsigned int		bi_bvec_done;   /* number of bytes completed in
+                                                  current bvec */
+};
+
 /*
  * main unit of I/O for the block layer and lower layers (ie drivers and
  * stacking drivers)
@@ -45,6 +56,8 @@ struct bio {
 	unsigned short		bi_vcnt;	/* how many bio_vec's */
 	unsigned short		bi_idx;		/* current index into bvl_vec */
 
+	struct bvec_iter	bi_iter;
+
 	/* Number of segments in this BIO after
 	 * physical address coalescing is performed.
 	 */
