--- a/drivers/mxc/hdmi-cec/mxc_hdmi-cec.c
+++ b/drivers/mxc/hdmi-cec/mxc_hdmi-cec.c
@@ -469,7 +403,6 @@ static long hdmi_cec_ioctl(struct file *filp, u_int cmd,
 		break;
 
 	case HDMICEC_IOC_STOPDEVICE:
-		hdmi_writeb(0x10, HDMI_CEC_CTRL);
 
 		val = HDMI_IH_CEC_STAT0_WAKEUP | HDMI_IH_CEC_STAT0_ERROR_FOLL |
 			HDMI_IH_CEC_STAT0_ERROR_INIT | HDMI_IH_CEC_STAT0_ARB_LOST |
@@ -480,6 +413,8 @@ static long hdmi_cec_ioctl(struct file *filp, u_int cmd,
 
 		hdmi_writeb(0x0, HDMI_CEC_POLARITY);
 
+		hdmi_writeb(0x10, HDMI_CEC_CTRL);
+
 		val = hdmi_readb(HDMI_MC_CLKDIS);
 		val |= HDMI_MC_CLKDIS_CECCLK_DISABLE;
 		hdmi_writeb(val, HDMI_MC_CLKDIS);
