commit 3b2a638bd8bb011037490b181f4aa2036e63190a
Author: Matus Kral <matuskral@me.com>
Date:   Tue Aug 5 22:26:02 2014 +0200

    very minimal DVI support - with EDID read

diff --git a/drivers/video/mxc/mxc_hdmi.c b/drivers/video/mxc/mxc_hdmi.c
index d0f31fa..3ee8205 100644
--- a/drivers/video/mxc/mxc_hdmi.c
+++ b/drivers/video/mxc/mxc_hdmi.c
@@ -1586,8 +1586,13 @@ static int mxc_edid_read_internal(struct mxc_hdmi *hdmi, unsigned char *edid,
 
 	ret = mxc_edid_parse_ext_blk(edid + EDID_LENGTH,
 			cfg, &fbi->monspecs);
-	if (ret < 0)
-		return -ENOENT;
+	if (ret < 0) {
+		fb_edid_add_monspecs(edid + EDID_LENGTH, &fbi->monspecs);
+		if (fbi->monspecs.modedb_len > 0)
+			hdmi->edid_cfg.hdmi_cap = false;
+		else
+			return -ENOENT;
+	}
 
 	/* need read segment block? */
 	if (extblknum > 1) {
@@ -1810,7 +1815,8 @@ static void mxc_hdmi_edid_rebuild_modelist(struct mxc_hdmi *hdmi)
 		 */
 		mode = &hdmi->fbi->monspecs.modedb[i];
 
-		if (!(mode->vmode & FB_VMODE_INTERLACED)) {
+		if (!(mode->vmode & FB_VMODE_INTERLACED) ||
+		     mxc_edid_mode_to_vic(mode) == 0) {
 			int vic = mxc_edid_mode_to_vic(mode);
 
 			dev_dbg(&hdmi->pdev->dev, "%s: Added mode %d(VIC %u):", __func__, i, vic);
diff --git a/drivers/video/mxc/mxc_ipuv3_fb.c b/drivers/video/mxc/mxc_ipuv3_fb.c
index 0e80e3f..49abe8c 100644
--- a/drivers/video/mxc/mxc_ipuv3_fb.c
+++ b/drivers/video/mxc/mxc_ipuv3_fb.c
@@ -2345,7 +2345,6 @@ static int mxcfb_probe(struct platform_device *pdev)
 	struct fb_info *fbi;
 	struct mxcfb_info *mxcfbi;
 	struct resource *res;
-	char *fix = kmalloc(11, GFP_KERNEL);
 	int ret = 0;
 
 	dev_dbg(&pdev->dev, "%s enter\n", __func__);
@@ -2355,11 +2354,6 @@ static int mxcfb_probe(struct platform_device *pdev)
 		return pdev->id;
 	}
 
-	strlcpy(fix, pdev->name, 11);
-	pr_info("FB name %s\n", pdev->name);
-	pr_info("FB name %s\n", fix);
-	pdev->name = fix;
-
 	plat_data = devm_kzalloc(&pdev->dev, sizeof(struct
 					ipuv3_fb_platform_data), GFP_KERNEL);
 	if (!plat_data)
diff --git a/sound/soc/fsl/fsl_hdmi.c b/sound/soc/fsl/fsl_hdmi.c
index fb9e852..f190e8c 100644
--- a/sound/soc/fsl/fsl_hdmi.c
+++ b/sound/soc/fsl/fsl_hdmi.c
@@ -338,6 +338,9 @@ static int fsl_hdmi_update_constraints(struct snd_pcm_substream *substream)
 
 	hdmi_get_edid_cfg(&edid_cfg);
 
+	if (!edid_cfg.hdmi_cap)
+		return -1;
+
 	fsl_hdmi_get_playback_rates();
 	ret = snd_pcm_hw_constraint_list(runtime, 0, SNDRV_PCM_HW_PARAM_RATE,
 			&playback_constraint_rates);
@@ -369,6 +372,10 @@ static int fsl_hdmi_soc_startup(struct snd_pcm_substream *substream,
 	struct imx_hdmi *hdmi_data = snd_soc_dai_get_drvdata(dai);
 	int ret;
 
+	ret = fsl_hdmi_update_constraints(substream);
+	if (ret < 0)
+		return ret;
+
 	clk_prepare_enable(hdmi_data->isfr_clk);
 	clk_prepare_enable(hdmi_data->iahb_clk);
 
@@ -376,10 +383,6 @@ static int fsl_hdmi_soc_startup(struct snd_pcm_substream *substream,
 			(int)clk_get_rate(hdmi_data->isfr_clk),
 			(int)clk_get_rate(hdmi_data->iahb_clk));
 
-	ret = fsl_hdmi_update_constraints(substream);
-	if (ret < 0)
-		return ret;
-
 	/* Indicates the subpacket represents a flatline sample */
 	hdmi_audio_writeb(FC_AUDSCONF, AUD_PACKET_SAMPFIT, 0x0);
 
diff --git a/sound/soc/fsl/imx-spdif.c b/sound/soc/fsl/imx-spdif.c
index 0b9d2d7..c005b0e 100644
--- a/sound/soc/fsl/imx-spdif.c
+++ b/sound/soc/fsl/imx-spdif.c
@@ -116,13 +116,13 @@ static int imx_spdif_audio_remove(struct platform_device *pdev)
 	struct snd_soc_card *card = platform_get_drvdata(pdev);
 	struct imx_spdif_data *data = snd_soc_card_get_drvdata(card);
 
+	snd_soc_unregister_card(card);
+
 	if (data->rxdev)
 		platform_device_unregister(data->rxdev);
 	if (data->txdev)
 		platform_device_unregister(data->txdev);
 
-	snd_soc_unregister_card(card);
-
 	return 0;
 }
 
